# Introduction
In this project I will be looking at 3 different datasets containing information on oil well sites. Each datasets pertains to a different region. My goal will be to determine which region has the highest profit margin. I will do so by bulding and training a model for each region to help predict which region has the highest profit margin. I will then use the bootstrapping technique to analyze potential risks of loss, as well as potential profits.

**Data Description** \
id - unique oil well identifier \
f0, f1, f2 - three features of points (their specific meaning is unimportant,
but the features themselves are significant) \
product - volume of reserves in the oil well (thousand barrels).



In [1]:
#importing all necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

## Download and prepare data

In [2]:
g1 = pd.read_csv('/datasets/geo_data_0.csv')
g2 = pd.read_csv('/datasets/geo_data_1.csv')
g3 = pd.read_csv('/datasets/geo_data_2.csv')

### first region (g1)

In [3]:
g1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
g1.sample(10)

id        f0        f1        f2     product
71096  0OGtB -0.376766  0.997739 -1.151282   87.726259
25409  VrcKF -0.957874  0.662481  1.315159   22.382423
93886  3JSoB  2.181795  0.317513  0.969483   22.972102
86388  X07nS  1.693292 -0.032616  1.825802  131.227417
62561  jr1BF  0.881963 -0.597878  4.243335  151.698842
20828  ZdCcS  0.568894 -0.582216  3.788057   83.006789
32894  uyaH5  0.057486  0.315614  2.438243  137.280633
38789  0ZLqu  1.024039 -0.578641  4.569445  125.168362
75724  2JepT  0.316141  0.777911  5.252772  133.009301
36545  1sZ5N -0.761979  0.623308  1.272795  118.807370

In [5]:
g1.duplicated().sum()

0

Ensuring there's no duplicates in the dataset.

In [6]:
g1 = g1.drop(['id'], axis=1)
g1.sample(5)

f0        f1        f2     product
71840  1.535694 -0.180294  1.136118  157.635840
84518 -0.855466  0.702869 -3.825249   10.175188
52245  0.846238 -0.605206  6.101787  158.513536
65262 -0.105974  0.533061 -0.451872   80.336055
7148  -0.200266  1.172046 -0.397995   29.150505

Removing any unnecessary columns, in this case just the id column.

### second region (g2)

In [7]:
g2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [8]:
g2.sample(10)

id         f0         f1        f2     product
97293  vUj4E  -2.867097  -2.723789  3.998160  110.992147
18446  ygdhE   9.762524  -6.545321  4.992791  134.766305
26599  7i1JH  10.607520  -8.048492  5.010137  134.766305
6765   gmG0j   2.706365  -8.420295  2.004481   53.906522
38241  z6JVq   6.718611  -3.302086  2.004947   53.906522
99497  HLInW   6.364759  -6.816702  5.005193  134.766305
84804  3HuEW  -5.928577 -11.295744  3.996052  110.992147
26638  nbHne   3.323624  -9.117453  0.001284    3.179103
9003   q4MtN -14.348000  -1.182799  3.994257  110.992147
79772  BikQS  -2.094906 -11.497634  2.998381   84.038886

In [9]:
g2.duplicated().sum()

0

Ensuring there are no duplicates in the dataset. 

In [10]:
g2 = g2.drop(['id'], axis=1)
g2.sample(5)

f0         f1        f2     product
7369   1.102415 -13.860526  0.999889   30.132364
20925 -9.348770 -12.144895  3.002714   84.038886
34783  4.889770  -1.662401  4.997403  137.945408
9990  -4.286693  -0.868957  2.010357   57.085625
18907  3.619001  -3.449466  4.001484  107.813044

Removing any unnecessary columns, in this case just the id column.

### third region (g3)

In [11]:
g3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [12]:
g3.duplicated().sum()

0

Ensuring that there are no duplicates in the dataset. 

In [13]:
g3.sample(10)

id        f0        f1        f2     product
6115   QmhJW  0.614516 -0.077983 -2.388863   53.538275
29979  trk0i -3.648740  0.644784  4.624971   67.177934
77494  BGHCR  0.835793 -3.522075 -2.063421   59.025245
11335  d19LU  1.319483 -2.674634  4.181363  152.053493
53316  OgPSF -2.096880 -1.534500 -5.334918   45.387704
73525  eNdnL  0.640213  1.867471  0.891576   57.325672
64113  k8g9N -0.230104 -1.009412  1.172677   95.787002
13144  5rG9v -1.035297 -2.692639  2.934152  145.089187
93535  AChyd  0.690154  0.060484 -2.118324   93.762183
31233  GIG8S  2.689870 -0.874373  6.077201  154.575737

In [14]:
g3 = g3.drop(['id'], axis=1)
g3.sample(5)

f0        f1        f2     product
71033  0.990106 -0.620968  4.753109  145.847613
8587  -1.341940  1.646953 -2.822217  107.247892
7637   0.139644  0.176932  5.499236   76.820234
16785  0.104135  1.148835  9.269552   70.476335
11131  3.929802 -1.850653  0.759533   73.516348

Removing any unnecessary columns, in this case just the id column.

## Train and test model for each region 

### split data into training and validation sets, ratio 75:25

In [15]:
#first region(g1)
g1_train, g1_valid = train_test_split(g1, test_size=0.25, random_state=315)

In [16]:
#second region(g2)
g2_train, g2_valid = train_test_split(g2, test_size=0.25, random_state=315)

In [17]:
#third region(g3)
g3_train, g3_valid = train_test_split(g3, test_size=0.25, random_state=315)

In [18]:
#training variables for g1
g1_train_features = g1_train.drop(['product'],axis=1)
g1_train_target = g1_train['product']
#validation variables for g1
g1_valid_features = g1_valid.drop(['product'], axis=1)
g1_valid_target = g1_valid['product']

In [19]:
print(g1_train_features.shape)
print(g1_train_target.shape)
print(g1_valid_features.shape)
print(g1_valid_target.shape)

(75000, 3)
(75000,)
(25000, 3)
(25000,)


In [20]:
#training variables for g2
g2_train_features = g2_train.drop(['product'],axis=1)
g2_train_target = g2_train['product']
#validation variables for g2
g2_valid_features = g2_valid.drop(['product'], axis=1)
g2_valid_target = g2_valid['product']

In [21]:
print(g2_train_features.shape)
print(g2_train_target.shape)
print(g2_valid_features.shape)
print(g2_valid_target.shape)

(75000, 3)
(75000,)
(25000, 3)
(25000,)


In [22]:
#training variables for g3
g3_train_features = g3_train.drop(['product'],axis=1)
g3_train_target = g3_train['product']
#validation variables for g3
g3_valid_features = g3_valid.drop(['product'], axis=1)
g3_valid_target = g3_valid['product']

In [23]:
print(g3_train_features.shape)
print(g3_train_target.shape)
print(g3_valid_features.shape)
print(g3_valid_target.shape)

(75000, 3)
(75000,)
(25000, 3)
(25000,)


### g1 model 

In [24]:
g1_model = LinearRegression()
g1_model.fit(g1_train_features, g1_train_target)
predictions1 = g1_model.predict(g1_valid_features)
print('Average volume of predicted reserves in g1_model:',predictions1.mean())

result = mean_squared_error(g1_valid_target, predictions1)**0.5
print('RMSE of g1 model is:',result)

Average volume of predicted reserves in g1_model: 92.55212915166508
RMSE of g1 model is: 37.91252286982915


### g2 model

In [25]:
g2_model = LinearRegression()
g2_model.fit(g2_train_features, g2_train_target)
predictions2 = g2_model.predict(g2_valid_features)
print('Average volume of predicted reserves in g2 model:',predictions2.mean())

result = mean_squared_error(g2_valid_target, predictions2)**0.5
print('RMSE of g2 model is:',result)

Average volume of predicted reserves in g2 model: 68.46363398543849
RMSE of g2 model is: 0.8907712786831072


### g3 model

In [26]:
g3_model = LinearRegression()
g3_model.fit(g3_train_features, g3_train_target)
predictions3 = g3_model.predict(g3_valid_features)
print('Average volume of predicted reserves for g3 model:',predictions3.mean())

result = mean_squared_error(g3_valid_target, predictions3)**0.5
print('RMSE of g3 model is:', result)

Average volume of predicted reserves for g3 model: 94.8301814760698
RMSE of g3 model is: 40.12203228780136


### Analyze results


In [27]:
print(g1['product'].describe())
print(g2['product'].describe())
print(g3['product'].describe())

count    100000.000000
mean         92.500000
std          44.288691
min           0.000000
25%          56.497507
50%          91.849972
75%         128.564089
max         185.364347
Name: product, dtype: float64
count    100000.000000
mean         68.825000
std          45.944423
min           0.000000
25%          26.953261
50%          57.085625
75%         107.813044
max         137.945408
Name: product, dtype: float64
count    100000.000000
mean         95.000000
std          44.749921
min           0.000000
25%          59.450441
50%          94.925613
75%         130.595027
max         190.029838
Name: product, dtype: float64


Based on the RMSE values for each region's model, it seems that region 2 is by far the easiest to predict. Region 2 has a very low RMSE value of 0.89, whereas region 1 has 37.91, and region 3 with 40.12. This would suggest that the data is more varied in regions 1 and 3, and that they are harder to predict. With that being said, the average volume of reserves is higher in both regions 1 (92.55) and 3 (94.83) than region 2 (68.46). This would suggest that regions 1 and 3 have a higher chance of producing the highest total profit. We will continue on and find out. 

## Prepare for profit calculation

### Store key values in variables

In [28]:
budget = 100_000_000 
wells = 200
rev_per_unit = 4500

In [29]:
g1_avg = g1['product'].mean()
g2_avg = g2['product'].mean()
g3_avg = g3['product'].mean()

### Calculate volume of reserves needed to breakeven

In [30]:
#cost per well budget
well_cost = budget/wells
print('Cost per well in USD is:', well_cost)

Cost per well in USD is: 500000.0


In [31]:
vol_needed = well_cost/rev_per_unit
print('Volume needed to breakeven is:',vol_needed)
print('Average volume of region 1 is:', g1_avg)
print('Average volume of region 2 is:', g2_avg)
print('Average volume of region 3 is:', g3_avg)

Volume needed to breakeven is: 111.11111111111111
Average volume of region 1 is: 92.50000000000001
Average volume of region 2 is: 68.82500000000002
Average volume of region 3 is: 95.00000000000004


### Findings
Based on the calculation of the  volume needed to breakeven, none of the 3 regions have an average volume sufficient enough to breakeven. The averages of regions 1 and 3 are much closer to the necessary volume than the average from region 2. This would suggest that the best options for profitability would be either region 1 or region 3. 

## Write a function to calculate profit

In [32]:
def calculate_profit(target, predictions):
    predictions = pd.Series(predictions, index=target.index)
    sorted_indices = predictions.sort_values(ascending=False).head(wells).index
    top_wells = target.loc[sorted_indices]
    target_reserves = top_wells.sum() 
    revenue = target_reserves * rev_per_unit
    profit = revenue - budget
    
    return profit.round(2)

In [33]:
g1_valid_target = g1_valid_target.reset_index(drop=True)
g2_valid_target = g2_valid_target.reset_index(drop=True)
g3_valid_target = g3_valid_target.reset_index(drop=True)

print(f'region g1 top 200 wells predicted profit:{calculate_profit(g1_valid_target, predictions1):,}')
print(f'region g2 top 200 wells predicted profit:{calculate_profit(g2_valid_target, predictions2):,}')
print(f'region g3 top 200 wells predicted profit:{calculate_profit(g3_valid_target, predictions3):,}')

region g1 top 200 wells predicted profit:33,481,224.4
region g2 top 200 wells predicted profit:24,150,866.97
region g3 top 200 wells predicted profit:26,280,301.64


Based on the predicted profits for the top 200 wells of each region, the region that I would suggest using for oil well development is g1. The predicted profits of this region are 6,000,000 more than the region with the second highest predicted profit. 

## Bootstrap Analysis

In [34]:
state = np.random.RandomState(315)
def bootstrap_profit(target, predictions, n_iterations=1000, sample_size=500):
    profits = []
   
    predictions = pd.Series(predictions, index=target.index)
    
    for i in range(n_iterations):
        sample_indices = state.choice(target.index, size=sample_size, replace=True)
        sample_predictions = predictions[sample_indices]
        sample_targets = target.loc[sample_indices]

        profit = calculate_profit(sample_targets, sample_predictions)
        profits.append(profit)
    
    profits = np.array(profits)
    mean_profit = profits.mean()
    confidence_interval = np.percentile(profits, [2.5, 97.5]).round(2)
    risk_of_loss = (profits < 0).mean()
    
    
    return mean_profit, confidence_interval, risk_of_loss


# Apply to each region
mean_profit_g1, ci_g1, risk_g1 = bootstrap_profit(g1_valid_target, predictions1)
mean_profit_g2, ci_g2, risk_g2 = bootstrap_profit(g2_valid_target, predictions2)
mean_profit_g3, ci_g3, risk_g3 = bootstrap_profit(g3_valid_target, predictions3)
# Print results
print(f"Region g1: Mean Profit: {mean_profit_g1.round(2):,}, 95% CI: {ci_g1}, Risk of Loss: {risk_g1*100:.2f}%")
print(f"Region g2: Mean Profit: {mean_profit_g2.round(2):,}, 95% CI: {ci_g2}, Risk of Loss: {risk_g2*100:.2f}%")
print(f"Region g3: Mean Profit: {mean_profit_g3.round(2):,}, 95% CI: {ci_g3}, Risk of Loss: {risk_g3*100:.2f}%")

Region g1: Mean Profit: 6,069,933.94, 95% CI: [  167480.3  12318968.32], Risk of Loss: 2.00%
Region g2: Mean Profit: 6,447,007.45, 95% CI: [ 1323963.6  11684605.36], Risk of Loss: 0.70%
Region g3: Mean Profit: 5,325,393.63, 95% CI: [ -993032.94 12254004.69], Risk of Loss: 5.30%


## Conclusion
After conducting the Bootstrap analysis, it seems that region g2 is actually the best region for oil well development. We see that the risk of loss in this region is only 0.7%, whereas regions g1 and g3 have 2% and 5.3% respectively. Region g2 sees the highest mean profits of the 3 regions with \$6,447,007,45. It should be noted that if you look at the confidence interval results, both regions g1 and g3 have a higher profit total in the 97.5 percentile. This could lend to why both of these regions saw higher profit total than region g2 in section 4. With that said, the 2.5 percentile for region g2 is significantly higher than both other regions. This can be supported by region g2 having a very low RMSE score. This was the easiest region to predict, and in turn would be my choice for the most profitable region to develop new oil well sites. 